lib/src/config_fetcher.dart

```diff
@@ -1,82 +1,210 @@
+    implements Fetcher {
+  static const globalBaseUrl = 'https://cdn-global.configcat.com';
+  static const euOnlyBaseUrl = 'https://cdn-eu.configcat.com';
+  static const _userAgentHeaderName = 'X-ConfigCat-UserAgent';
+  static const _ifNoneMatchHeaderName = 'If-None-Match';
+  static const _etagHeaderName = 'Etag';
+  static const _successStatusCodes = [200, 201, 202, 203, 204];
+
+  late final ConfigCatLogger _logger;
+  late final ConfigJsonCache _jsonCache;
+  late final String _mode;
+  late final String _sdkKey;
+
+  late final bool _urlIsCustom;
+  late final Dio _client;
+  late String _etag = '';
+  late String _url;
+
+  ConfigFetcher(
+      {required ConfigCatLogger logger,
+      required String sdkKey,
+      required String mode,
+      required ConfigJsonCache jsonCache,
+      required ConfigCatOptions options}) {
+    _logger = logger;
+    _jsonCache = jsonCache;
+    _mode = mode;
+    _sdkKey = sdkKey;
+
+    _urlIsCustom = options.baseUrl.isNotEmpty;
+    _url = _urlIsCustom
+        ? options.baseUrl
+        : options.dataGovernance == DataGovernance.global
+            ? globalBaseUrl
+            : euOnlyBaseUrl;
+
+    _client = Dio(BaseOptions(
+        connectTimeout: options.connectTimeout.inMilliseconds,
+        receiveTimeout: options.receiveTimeout.inMilliseconds,
+        sendTimeout: options.sendTimeout.inMilliseconds,
+        responseType: ResponseType.plain,
+        validateStatus: (status) =>
+            status != null && (status >= 200 && status < 600)));
+
+    if (options.httpClientAdapter != null) {
+      _client.httpClientAdapter = options.httpClientAdapter!;
+    }
+
+    if (options.proxyUrl.isNotEmpty &&
+        _client.httpClientAdapter is DefaultHttpClientAdapter) {
+      (_client.httpClientAdapter as DefaultHttpClientAdapter)
+          .onHttpClientCreate = (client) {
+        client.findProxy = (uri) {
+          return 'PROXY ${options.proxyUrl}';
+        };
+      };
+    }
   }
 
-  Future<FetchResponse> fetchConfigurationJson(http.Client client) {
-    return _executeFetch(client, executionCount: 2);
+  /// Gets the underlying [Dio] client.
+  @override
+  Dio get client {
+    return _client;
   }
 
-  Future<FetchResponse> _executeFetch(http.Client client, {int executionCount = 1}) async {
-    Map<String, String> headers = {
-      'X-ConfigCat-UserAgent': 'ConfigCat-Dart/$mode-$_version',
-    };
+  /// Fetches the current ConfigCat configuration json.
+  @override
+  Future<FetchResponse> fetchConfiguration() {
+    // ensure fetch requests are not initiated simultaneously
+    return syncFuture(() => _executeFetch(2));
+  }
+
+  /// Closes the underlying http connection.
+  @override
+  void close() {
+    _client.close();
+  }
+
+  Future<FetchResponse> _executeFetch(int executionCount) async {
+    final response = await _doFetch();
+    if (!response.isFetched || response.config.preferences == null) {
+      return response;
+    }
```

Comment:
We can remove this part, right? The code below already does the same thing

Reply:
Yep, overlooked this part :)